### PR TITLE
Add webserver layer: gunicorn + nginx.

### PR DESCRIPTION
**Pull request by iquintero**
this includes a 'serve' cli command that can be used instead of the old
style entrypoint that handles both training and serving.

*Issue #, if available:*

*Description of changes:*

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: etc/nginx.conf [hunk 1]
```diff
@@ -0,0 +1,38 @@
...
+error_log  /dev/stderr;
+
+worker_rlimit_nofile 4096;
+
+events {
+  worker_connections 2048;
+}
+
+http {
+  # include /etc/nginx/mime.types;
```
Comment (mvsusp) [inline etc/nginx.conf]: Did you notice that master is different: https://github.com/aws/sagemaker-containers/blob/master/src/container_support/etc/nginx.conf#L13

File: etc/nginx.conf [hunk 2]
```diff
@@ -0,0 +1,38 @@
+worker_processes auto;
+daemon off;
+pid /tmp/nginx.pid;
+error_log  /dev/stderr;
```
Comment (mvsusp) [inline etc/nginx.conf]: Did you notice that master is different: https://github.com/aws/sagemaker-containers/blob/master/src/container_support/etc/nginx.conf#L4

File: src/sagemaker_containers/cli/serve.py [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+# may not use this file except in compliance with the License. A copy of
+# the License is located at
+#
+#     http://aws.amazon.com/apache2.0/
+#
+# or in the 'license' file accompanying this file. This file is
+# distributed on an 'AS IS' BASIS, WITHOUT WARRANTIES OR CONDITIONS OF
+# ANY KIND, either express or implied. See the License for the specific
+# language governing permissions and limitations under the License.
+import sagemaker_containers
```
Comment (mvsusp) [inline src/sagemaker_containers/cli/serve.py]: nitpick: sagemaker_containers, smc, or something else. Let's decide this before migrating. Do not need to be here.

File: src/sagemaker_containers/cli/serve.py [hunk 4]
```diff
@@ -0,0 +1,22 @@
...
+
+
+def main():
+    env = sagemaker_containers.environment.ServingEnvironment()
+    sagemaker_containers.server.start_server(env.flask_app)
```
Comment (mvsusp) [inline src/sagemaker_containers/cli/serve.py]: I would rename flask_app to framework_module or app_module.

The reason is that gunicorn runs more things than just Flask http://docs.gunicorn.org/en/stable/run.html

File: src/sagemaker_containers/environment.py [hunk 5]
```diff
@@ -73,6 +73,7 @@
 MODEL_SERVER_WORKERS_ENV = 'SAGEMAKER_MODEL_SERVER_WORKERS'  # type: str
 MODEL_SERVER_TIMEOUT_ENV = 'SAGEMAKER_MODEL_SERVER_TIMEOUT'  # type: str
 USE_NGINX_ENV = 'SAGEMAKER_USE_NGINX'  # type: str
+FLASK_APP_ENV = 'SAGEMAKER_FLASK_APP'  # type: str
```
Comment (mvsusp) [inline src/sagemaker_containers/environment.py]: Look observation above.

File: src/sagemaker_containers/server.py [hunk 6]
```diff
@@ -0,0 +1,63 @@
...
+    def terminate(signal_number, stack_frame):
+        process.terminate()
+
+    signal.signal(signal.SIGTERM, terminate)
+
+
+def start_server(module_app):
+
+    env = sagemaker_containers.environment.ServingEnvironment()
+    gunicorn_bind_address = '0.0.0.0:8080'
```
Comment (mvsusp) [inline src/sagemaker_containers/server.py]: nitpick: constant

File: src/sagemaker_containers/server.py [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+
+    nginx = None
+
+    if env.use_nginx:
+        gunicorn_bind_address = 'unix:/tmp/gunicorn.sock'
+        nginx = subprocess.Popen(['nginx', '-c', '/tmp/nginx.conf'])
```
Comment (mvsusp) [inline src/sagemaker_containers/server.py]: Do we need access and error logs: https://github.com/aws/sagemaker-containers/blob/master/src/container_support/serving.py#L87

File: src/sagemaker_containers/server.py [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+
+        add_terminate_signal(nginx)
+
+    gunicorn = subprocess.Popen(['gunicorn',
+                                 '--timeout', str(env.model_server_timeout),
+                                 '-k', 'gevent',
+                                 '-b', gunicorn_bind_address,
+                                 '--worker-connections', str(1000 * env.model_server_workers),
+                                 '-w', str(env.model_server_workers),
+                                 '--log-level', 'debug',
```
Comment (mvsusp) [inline src/sagemaker_containers/server.py]: Consider setting it to info

File: test/functional/test_server.py [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+import sagemaker_containers
+
+
+def test_server():
+    os.environ[sagemaker_containers.environment.FLASK_APP_ENV] = 'test.functional.simple_flask:app'
+    os.environ[sagemaker_containers.environment.USE_NGINX_ENV] = 'false'
+
+    env = sagemaker_containers.environment.ServingEnvironment()
+    print(env.flask_app)
```
Comment (mvsusp) [inline test/functional/test_server.py]: remove the print

File: src/sagemaker_containers/server.py [hunk 10]
```diff
@@ -0,0 +1,63 @@
...
+
+
+def add_terminate_signal(process):
+    def terminate(signal_number, stack_frame):
+        process.terminate()
+
+    signal.signal(signal.SIGTERM, terminate)
+
+
+def start_server(module_app):
```
Comment (mvsusp) [inline src/sagemaker_containers/server.py]: What about just start?

File: test/unit/test_serve_cli.py [hunk 11]
```diff
@@ -0,0 +1,24 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
```
Comment (mvsusp) [inline test/unit/test_serve_cli.py]: what about keeping the same file structure of src: test/unit/cli/test_serve.py

File: test/unit/test_server.py [hunk 12]
```diff
@@ -0,0 +1,65 @@
...
+# language governing permissions and limitations under the License.
+from mock import call, patch, PropertyMock
+
+from sagemaker_containers.environment import ServingEnvironment
+from sagemaker_containers.server import start_server
+
+
+@patch.object(ServingEnvironment, 'model_server_workers', new_callable=PropertyMock)
+@patch.object(ServingEnvironment, 'model_server_timeout', new_callable=PropertyMock)
+@patch.object(ServingEnvironment, 'use_nginx', new_callable=PropertyMock)
```
Comment (mvsusp) [inline test/unit/test_server.py]: tip = PropertyMock(return_value=False)

File: etc/nginx.conf (hunk 2 above)
Comment (iquintero) [reply etc/nginx.conf]: this one is on purpose, it avoids doing the symlink shenanigans to get the error log to stderr.

File: src/sagemaker_containers/cli/serve.py (hunk 4 above)
Comment (iquintero) [reply src/sagemaker_containers/cli/serve.py]: will do

File: etc/nginx.conf (hunk 2 above)
Comment (mvsusp) [reply etc/nginx.conf]: 👍

**Pull request merged**